storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java

```diff
@@ -0,0 +1,95 @@
+package com.pushtorefresh.storio;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Common class for search and cache indirect type mapping.
+ */
+public class TypeMappingSearcher {
+
+    @Nullable
+    private final Map<Class<?>, ? extends TypeMapping<?>> directTypesMapping;
+
+    @NonNull
+    private final Map<Class<?>, TypeMapping<?>> indirectTypesMappingCache
+            = new ConcurrentHashMap<Class<?>, TypeMapping<?>>();
+
+    public TypeMappingSearcher(@Nullable Map<Class<?>, ? extends TypeMapping<?>> directTypesMapping) {
+        this.directTypesMapping = directTypesMapping;
+    }
+
+    /**
+     * Search algorithm:
```

Comment:
Let's move description of algorithm to the comment inside of the method, it's not part of documentation :)